drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -35,1403 +35,1532 @@
-}
-
-void RasterizerCanvasGLES2::canvas_begin() {
 
-	state.canvas_shader.bind();
-	state.using_transparent_rt = false;
-	int viewport_x, viewport_y, viewport_width, viewport_height;
+	// pushing back from local variable .. not ideal but has to use a Vector because non pod
+	// due to RIDs
+	BatchTex b;
+	b.RID_texture = p_texture;
+	b.RID_normal = p_normal;
 
-	if (storage->frame.current_rt) {
-		glBindFramebuffer(GL_FRAMEBUFFER, storage->frame.current_rt->fbo);
-		state.using_transparent_rt = storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT];
+	// get the texture
+	RasterizerStorageGLES2::Texture *texture = _get_canvas_texture(p_texture);
 
-		if (storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_DIRECT_TO_SCREEN]) {
-			// set Viewport and Scissor when rendering directly to screen
-			viewport_width = storage->frame.current_rt->width;
-			viewport_height = storage->frame.current_rt->height;
-			viewport_x = storage->frame.current_rt->x;
-			viewport_y = OS::get_singleton()->get_window_size().height - viewport_height - storage->frame.current_rt->y;
-			glScissor(viewport_x, viewport_y, viewport_width, viewport_height);
-			glViewport(viewport_x, viewport_y, viewport_width, viewport_height);
-			glEnable(GL_SCISSOR_TEST);
-		}
+	if (texture) {
+		b.tex_pixel_size.x = 1.0 / texture->width;
+		b.tex_pixel_size.y = 1.0 / texture->height;
+	} else {
+		// maybe doesn't need doing...
+		b.tex_pixel_size.x = 1.0;
+		b.tex_pixel_size.y = 1.0;
 	}
 
-	if (storage->frame.clear_request) {
-		glClearColor(storage->frame.clear_request_color.r,
-				storage->frame.clear_request_color.g,
-				storage->frame.clear_request_color.b,
-				state.using_transparent_rt ? storage->frame.clear_request_color.a : 1.0);
-		glClear(GL_COLOR_BUFFER_BIT);
-		storage->frame.clear_request = false;
-	}
+	if (p_tile) {
+		if (texture) {
+			// default
+			b.tile_mode = BatchTex::TILE_NORMAL;
 
-	/*
-	if (storage->frame.current_rt) {
-		glBindFramebuffer(GL_FRAMEBUFFER, storage->frame.current_rt->fbo);
-		glColorMask(1, 1, 1, 1);
+			// no hardware support for non power of 2 tiling
+			if (!storage->config.support_npot_repeat_mipmap) {
+				if (next_power_of_2(texture->alloc_width) != (unsigned int)texture->alloc_width && next_power_of_2(texture->alloc_height) != (unsigned int)texture->alloc_height) {
+					b.tile_mode = BatchTex::TILE_FORCE_REPEAT;
+				}
+			}
+		} else {
+			// this should not happen?
+			b.tile_mode = BatchTex::TILE_OFF;
+		}
+	} else {
+		b.tile_mode = BatchTex::TILE_OFF;
 	}
-	*/
 
-	reset_canvas();
+	// push back
+	bdata.batch_textures.push_back(b);
 
-	glActiveTexture(GL_TEXTURE0);
-	glBindTexture(GL_TEXTURE_2D, storage->resources.white_tex);
+	return bdata.batch_textures.size() - 1;
+}
 
-	glVertexAttrib4f(VS::ARRAY_COLOR, 1, 1, 1, 1);
-	glDisableVertexAttribArray(VS::ARRAY_COLOR);
+void RasterizerCanvasGLES2::_batch_upload_buffers() {
 
-	// set up default uniforms
+	// noop?
+	if (!bdata.vertices.size())
```

Comment:
Similar to the above, consider a WARN_PRINT or an ERR_PRINT here if appropriate

Reply:
This also can occur I believe in some situations where there is no error. Consider if it is only drawing a batch of defaults with no vertices in the vertex buffer, or the screen is blank. No need to upload.

It is unlikely I'll grant you though! :grin: